lib/FrameManager.js

```diff
@@ -243,12 +244,47 @@ class Frame {
    * @return {!Promise<(!Object|undefined)>}
    */
   async _rawEvaluate(pageFunction, ...args) {
-    const expression = helper.evaluationString(pageFunction, ...args);
-    const contextId = this._defaultContextId;
-    const { exceptionDetails, result: remoteObject }  = await this._client.send('Runtime.evaluate', { expression, contextId, returnByValue: false, awaitPromise: true});
+    if (helper.isString(pageFunction)) {
+      const contextId = this._defaultContextId;
+      const expression = pageFunction;
+      const { exceptionDetails, result: remoteObject } = await this._client.send('Runtime.evaluate', { expression, contextId, returnByValue: false, awaitPromise: true});
+      if (exceptionDetails)
+        throw new Error('Evaluation failed: ' + helper.getExceptionMessage(exceptionDetails));
+      return remoteObject;
+    }
+
+    const { exceptionDetails, result: remoteObject } = await this._client.send('Runtime.callFunctionOn', {
+      functionDeclaration: pageFunction.toString(),
+      executionContextId: this._defaultContextId,
+      arguments: args.map(convertArgument),
+      returnByValue: false,
+      awaitPromise: true
+    });
     if (exceptionDetails)
       throw new Error('Evaluation failed: ' + helper.getExceptionMessage(exceptionDetails));
     return remoteObject;
+
+    /**
+     * @param {*} arg
+     * @return {*}
+     */
+    function convertArgument(arg) {
+      if (Object.is(arg, -0))
+        return { unserializableValue: '-0' };
+      if (Object.is(arg, Infinity))
+        return { unserializableValue: 'Infinity' };
+      if (Object.is(arg, -Infinity))
+        return { unserializableValue: '-Infinity' };
+      if (Object.is(arg, NaN))
+        return { unserializableValue: 'NaN' };
+      if (arg instanceof ElementHandle) {
+        const objectId = arg._remoteObjectId();
+        if (!objectId)
```

Comment:
Let's check if all the execution contexts are the same.

Reply:
Done